File: frontend/src/types/api/order.ts
/** Single line item in an order (GET/POST response) */
export type OrderItem = {
  id: number;
  order_id: number;
  variant_id: number;
  sku: string;
  /** Unit price in integer minor units. */
  unit_price: number;
  currency: string;
  quantity: number;
  attributes: Record<string, string>;
  created_at: string;
};

/**
 * Order (GET array element or POST response).
 * Monetary fields (subtotal, shipping, discount, total, unit_price) are integer minor units.
 * `currency` is a financial snapshot — format with order.currency, not a later site currency.
 */
export type Order = {
  id: number;
  user_id: number;
  status: string;
  currency: string;
  subtotal: number;
  shipping: number;
  discount: number;
  total: number;
  items: OrderItem[];
  /** Present on POST create when a Stripe PaymentIntent is created */
  client_secret?: string;
};

/** Shipping address payload for PUT /orders/{id}/address */
export type OrderShippingAddressInput = {
  line1: string;
  line2?: string;
  city: string;
  region?: string;
  postal_code: string;
  country_code: string;
  phone?: string;
};

/** Saved order address (API response) */
export type OrderAddress = {
  id: number;
  order_id: number;
  type: string;
  phone?: string;
  line1: string;
  line2?: string;
  city: string;
  region?: string;
  postal_code: string;
  country_code: string;
  created_at: string;
};

File: backend/internal/model/order.go
package model

import (
	"encoding/json"
	"time"

	"alwis.dev/selectify/internal/types"
)

type Orders []Order

type Order struct {
	ID           uint              `db:"id" json:"id"`
	UserID       uint              `db:"user_id" json:"user_id"`
	Status       types.OrderStatus `db:"status" json:"status"`
	Currency     string            `db:"currency" json:"currency"`
	Subtotal     uint64            `db:"subtotal" json:"subtotal"`
	Shipping     uint64            `db:"shipping" json:"shipping"`
	Discount     uint64            `db:"discount" json:"discount"`
	Total        uint64            `db:"total" json:"total"`
	CreatedAt    time.Time         `db:"created_at" json:"-"`
	UpdatedAt    time.Time         `db:"updated_at" json:"-"`
	Items        OrderItems        `db:"-" json:"items,omitempty"`
	ClientSecret string            `db:"-" json:"client_secret,omitempty"`
}

type OrderItems []OrderItem

type OrderItem struct {
	ID         uint            `db:"id" json:"id"`
	OrderID    uint            `db:"order_id" json:"order_id"`
	VariantID  uint            `db:"variant_id" json:"variant_id"`
	SKU        string          `db:"sku" json:"sku"`
	UnitPrice  uint64          `db:"unit_price" json:"unit_price"`
	Quantity   uint            `db:"quantity" json:"quantity"`
	Attributes json.RawMessage `db:"attributes" json:"attributes"` // JSONB stored as string
	CreatedAt  time.Time       `db:"created_at" json:"created_at"`
}
